[#55270] Reduce visibility of all users when addings members to a project

oliverguenther opened the pull request:
https://community.openproject.org/projects/openproject/work_packages/55270/activity

oliverguenther committed:
Reduce visibility of principals to common memberships or view_all_principals permission

oliverguenther committed:
Invite users by email

oliverguenther committed:
Restrict visibility on groups scope

oliverguenther committed:
Disable RSpec/ChangeByZero to fix spec

The and chaining doesn't work with no change assumptions

oliverguenther committed:
Fix sharing inviting of users depending on project set

klaustopher commented on app/controllers/concerns/member_helper.rb:
```diff
@@ -56,21 +56,29 @@ def user_ids_for_new_members(member_params)
   def invite_new_users(user_ids, send_notification: true)
     user_ids.filter_map do |id|
       if id.present? && (id.to_i == 0 || EmailValidator.valid?(id)) # we've got an email - invite that user
```
Not related to this PR, but emails are allowed to start with numbers. So the `id.to_i` will return that number when the email starts with one.


```ruby
"123max.mustermann@example.com".to_i # => 123
```

Maybe we should only check for the EmailValidator?

klaustopher commented on app/models/groups/scopes/visible.rb:
```diff
@@ -33,12 +33,24 @@ module Visible
         else
-          Group
-            .in_project(Project.allowed_to(current_user, :view_members))
+          # Always use subqueries with unscoped to avoid join conflicts
+          where(
+            id: Group.unscoped.in_visible_project(user).select(:id)
+          ).or(
+            where(
+              id: Group.unscoped.containing_user(user).select(:id)
+            )
+          )
```
We have seen problems with `OR` in queries like this, because it is not properly optimized. Maybe try rewriting it using a `UNION ALL`

```ruby
where(Group.arel_table[:id].in(Arel::Nodes::UnionAll.new(
  Group.unscoped.containing_user(user).select(:id).arel,
  Group.unscoped.in_visible_project(user).select(:id).arel)
))
```

klaustopher commented on app/models/group.rb:
```diff
@@ -73,6 +73,10 @@ def self.synchronized_group_partials
 
   scopes :visible
 
+  def self.containing_user(user = User.current)
```
Why is this a class method and not a scope?

klaustopher commented on app/services/members/add_role_service.rb:
```diff
@@ -0,0 +1,53 @@
+      @current_user = current_user
+    end
+
+    def call(user_id:, role_id:, project_id:, send_notifications: true)
```
We have memberships on other resources than projects as well. Should we maybe have this service handle this as well?

klaustopher commented on app/services/members/remove_role_service.rb:
```diff
@@ -0,0 +1,54 @@
+      @current_user = current_user
+    end
+
+    def call(user_id:, role_id:, project_id:, send_notifications: true)
```
Same as above. Memberships can also be on entities

klaustopher commented on app/services/members/add_role_service.rb:
```diff
@@ -0,0 +1,53 @@
+    end
+
+    def call(user_id:, role_id:, project_id:, send_notifications: true)
+      member = Member.find_by(user_id:, project_id:)
```
Also, this could also return members on entites within a project. When we want to check for membership on the project itself, we need to check for `Member.find_by(user_id:, project_id:, entity: nil)`

klaustopher reviewed (changes requested):
Looks good to me in general. Some of the queries could be optimized a bit, could maybe be checked out.

What needs to be changed is the new services. At least we have to explicitly add `entity: nil` to the `find_by` method to exclude members of entities. Optimally we allow passing in an entity object and then we check

```ruby
member = if entity.is_a?(Project)
		   Member.find_by(project: entity, user: user)
		 else
		   Member.find_by(project: entity.project, user: user, entity: entity)
		 end
```

oliverguenther committed:
Return the invited user, even if send_notifications is suppressed

You invited the user after all, in case of sharing, we want a custom
notification

oliverguenther committed:
Use EmailValidator.valid?(id) instead of id.to_i

Remark from Klaus. Not related to this PR, but emails are allowed to
start with numbers. So the id.to_i will return that number when the
email starts with one.

oliverguenther committed:
Move containing_user into a scope and add spec

oliverguenther committed:
Fix new services to work/exclude entities

oliverguenther committed:
Try UNION on Group visible scope

oliverguenther commented:
@klaustopher the union seems to work fine. I addressed all your remarks 👍

klaustopher reviewed (commented):
Thanks! LGTM

oliverguenther committed:
Adapt CF validation after core changes
